File: apple/Tests/BidiLensTests/TokenBoundaryTests.swift
import XCTest
@testable import BidiLens

final class TokenBoundaryTests: XCTestCase {
    func testMathWhitespaceMatchesOtherCores() {
        for space in ["\u{feff}", "\u{a0}", "\u{202f}"] {
            for source in ["$\(space)x$", "$x\(space)$"] {
                XCTAssertEqual(BidiAnalyzer.findTechnicalTokenRanges(source).filter { $0.kind == .math }.map(\.text), [])
            }
        }
        for content in ["\u{85}", "\u{1c}"] {
            let source = "$\(content)x$"
            XCTAssertEqual(BidiAnalyzer.findTechnicalTokenRanges(source).filter { $0.kind == .math }.map(\.text), [source])
        }
    }

    func testAmountsRangesAndQuotesStayIntact() {
        for token in ["$10", "€12.50", "£25", "۱۰€", "10-20", "10–20", "-10--2", "۱۰-۲۰", "١٠-٢٠"] {
            let source = "👋 مقدار \(token) است."
            let ranges = BidiAnalyzer.analyze(source).isolations
            XCTAssertEqual(ranges.map(\.text), [token], token)
            if let range = ranges.first {
                XCTAssertEqual(UnicodeClassifier.substring(source, utf16Range: range.utf16Range), token)
            }
        }
        for quote in ["\"", "'", "“", "”", "«", "»"] {
            XCTAssertEqual(BidiAnalyzer.findTechnicalTokenRanges("مسیر \(quote)/usr/local/bin\(quote) است.").map(\.text), ["/usr/local/bin"])
        }
    }

    func testCurrencyProseAndEscapedDollarsAreNotMath() {
        for source in ["$ x$", "$x $", "$10 and $20", "\\$x\\$"] {
            XCTAssertEqual(BidiAnalyzer.findTechnicalTokenRanges(source).filter { $0.kind == .math }.map(\.text), [], source)
        }
        XCTAssertEqual(BidiAnalyzer.findTechnicalTokenRanges("هزینه این کتاب $10 و آن یکی $20 است.").map(\.text), ["$10", "$20"])
        for (source, expected) in [("قیمت $10 است و $x+1$ درست است.", "$x+1$"), ("$x\\$y$", "$x\\$y$")] {
            XCTAssertEqual(BidiAnalyzer.findTechnicalTokenRanges(source).filter { $0.kind == .math }.map(\.text), [expected])
        }
    }
}

File: windows/tests/BidiLens.Tests/TokenBoundaryTests.cs
using BidiLens;

internal static class TokenBoundaryTests
{
    internal static int Run()
    {
        var assertions = 0;
        void Equal(IEnumerable<string> actual, params string[] expected)
        {
            assertions++;
            if (!actual.SequenceEqual(expected)) throw new InvalidOperationException($"Token boundaries: {string.Join(", ", actual)} != {string.Join(", ", expected)}");
        }
        foreach (var token in new[] { "$10", "€12.50", "£25", "۱۰€", "10-20", "10–20", "-10--2", "۱۰-۲۰", "١٠-٢٠" })
        {
            var source = $"👋 مقدار {token} است.";
            var ranges = BidiAnalyzer.Analyze(source).Isolations;
            Equal(ranges.Select(range => range.Text), token);
            Equal(ranges.Select(range => source[range.Utf16Start..range.Utf16End]), token);
        }
        foreach (var quote in new[] { "\"", "'", "“", "”", "«", "»" })
            Equal(BidiAnalyzer.FindTechnicalTokenRanges($"مسیر {quote}/usr/local/bin{quote} است.").Select(range => range.Text), "/usr/local/bin");
        foreach (var source in new[] { "$ x$", "$x $", "$10 and $20", "\\$x\\$" })
            Equal(BidiAnalyzer.FindTechnicalTokenRanges(source).Where(range => range.Kind == TechnicalTokenKind.Math).Select(range => range.Text));
        foreach (var space in new[] { "\ufeff", "\u00a0", "\u202f" })
            foreach (var source in new[] { $"${space}x$", $"$x{space}$" })
                Equal(BidiAnalyzer.FindTechnicalTokenRanges(source).Where(range => range.Kind == TechnicalTokenKind.Math).Select(range => range.Text));
        foreach (var content in new[] { "\u0085", "\u001c" })
        {
            var source = $"${content}x$";
            Equal(BidiAnalyzer.FindTechnicalTokenRanges(source).Where(range => range.Kind == TechnicalTokenKind.Math).Select(range => range.Text), source);
        }
        Equal(BidiAnalyzer.FindTechnicalTokenRanges("هزینه این کتاب $10 و آن یکی $20 است.").Select(range => range.Text), "$10", "$20");
        foreach (var (source, expected) in new[] { ("قیمت $10 است و $x+1$ درست است.", "$x+1$"), ("$x\\$y$", "$x\\$y$") })
            Equal(BidiAnalyzer.FindTechnicalTokenRanges(source).Where(range => range.Kind == TechnicalTokenKind.Math).Select(range => range.Text), expected);
        return assertions;
    }
}
